### PR TITLE
SyncFlow: top level heartbeat

### PR DESCRIPTION
**Pull request by serprex**
was seeing heartbeat timeouts

### CODE REVIEW

**Comment by heavycrystal** [comment]
we lose more visibility into which phase of a batch we are in

**Comment by serprex** [comment]
yes, tho with parallelism heartbeats were already stepping over each other

**Comment by iskakaushik** [comment]
@serprex  - can we have this heartbeat contain a message that shows the consolidated status of syncs and normalizes, something like:

```
{
   start_sync_batch_id: ...,
   start_normalize_batch_id: ...,
   current_sync_batch_id: ...,
   current_sync_batch_start_time_utc: ...,
   current_normalize_batch_id: ...,
   current_normalize_start_time_utc: ...,
   current_time_utc: ...
}
```

**Comment by serprex** [comment]
updated with heartbeat state I think will be useful. Temporal already has timestamps on heartbeat, no need to put timestamps in message

File: flow/activities/flowable_core.go [hunk 1]
```diff
@@ -202,24 +200,25 @@ func syncCore[TPull connectors.CDCPullConnectorCore, TSync connectors.CDCSyncCon
...
 			if temporal.IsApplicationError(err) {
-				return 0, err
+				return nil, err
 			} else {
-				return 0, fmt.Errorf("failed in pull records when: %w", err)
+				return nil, fmt.Errorf("failed in pull records when: %w", err)
 			}
 		}
 		logger.Info("no records to push")
+		syncWaiting.Store(true)
```
Comment (heavycrystal) [inline flow/activities/flowable_core.go:209]: why would this be true? also unsure how we hit case in 208

File: flow/activities/flowable_core.go [hunk 2]
```diff
@@ -668,9 +672,11 @@ func (a *FlowableActivity) normalizeLoop(
 				break
 			}
 		case <-syncDone:
+			normalizeWaiting.Store(false)
 			logger.Info("[normalize-loop] syncDone closed")
 			return
 		case <-ctx.Done():
+			normalizeWaiting.Store(false)
```
Comment (heavycrystal) [inline flow/activities/flowable_core.go:679]: can we defer this instead?

File: flow/activities/flowable_core.go (hunk 1 above)
Comment (serprex) [reply flow/activities/flowable_core.go:209]: `Waiting` should generally be true as long as a sync/normalize isn't happening. Should probably initialize them as true. For normalize I settled on having that be true while channel is being waited on, with syncWaiting it's less clear, maybe a string saying what state it's in would be better, between setup, sync, schema deltas, & waiting on normalize

File: flow/activities/flowable_core.go (hunk 2 above)
Comment (serprex) [reply flow/activities/flowable_core.go:679]: Yes

File: flow/activities/flowable_core.go (hunk 1 above)
Comment (heavycrystal) [reply flow/activities/flowable_core.go:209]: sync being a string makes sense; or you could add an `Idle` state

File: flow/activities/flowable.go [hunk 3]
```diff
@@ -249,40 +242,67 @@ func (a *FlowableActivity) CreateNormalizedTable(
...
-) (CdcState, context.Context, error) {
+	options *protos.SyncFlowOptions,
+) error {
 	ctx = context.WithValue(ctx, shared.FlowNameKey, config.FlowJobName)
-	srcConn, err := connectors.GetByNameAs[connectors.CDCPullConnector](ctx, config.Env, a.CatalogPool, config.SourceName)
+	logger := activity.GetLogger(ctx)
+
+	var currentSyncFlowNum atomic.Int32
+	var totalRecordsSynced atomic.Int64
+	var normalizingBatchID atomic.Int64
```
Comment (heavycrystal) [inline flow/activities/flowable.go:255]: nit: could probably store some of these in substruct of `FlowableActivity` and keep calls clean

**Review by heavycrystal — APPROVED** [review]

**Pull request merged**